### PR TITLE
fix: collect leftover collateral after full auction raise

### PR DESCRIPTION
**Pull request by 0xJabberwock**
Closes HAI-329

### CODE REVIEW

File: src/contracts/proxies/actions/BasicActions.sol [hunk 1]
```diff
@@ -398,4 +398,26 @@ contract BasicActions is CommonActions, IBasicActions {
...
+  /// @inheritdoc IBasicActions
+  function collectSystemCoins(
+    address _manager,
+    address _coinJoin,
+    uint256 _safeId,
+    uint256 _deltaWad
+  ) external onlyDelegateCall {
+    // Transfers COIN amount to the user's address
+    _collectAndExitCoins(_manager, _coinJoin, _safeId, _deltaWad);
+  }
```
Comment (wei3erHase) [inline src/contracts/proxies/actions/BasicActions.sol]: not needed, we wanted to do minimal work here, but ok i'd say, i don't see a possibility where the safeHandler would get some coins tbh

File: src/interfaces/proxies/actions/ICommonActions.sol [hunk 2]
```diff
@@ -20,7 +20,7 @@ interface ICommonActions {
   /**
    * @notice Exits system coins from the safeEngine
    * @param  _coinJoin Address of the CoinJoin contract
-   * @param  _coinsToExit Amount of coins to exit [wad]
+   * @param  _coinsToExit Amount of coins to exit [wad] // REVIEW: _coinsToExit is treated as RAD!
    */
   function exitSystemCoins(address _coinJoin, uint256 _coinsToExit) external;
```
Comment (wei3erHase) [inline src/interfaces/proxies/actions/ICommonActions.sol]: it is! systemCoins are RAD.

File: test/scopes/ProxyUser.t.sol [hunk 3]
```diff
@@ -274,12 +274,26 @@ abstract contract ProxyUser is BaseUser, Contracts, ScriptBase {
     uint256 _coinsToExit = safeEngine.coinBalance(address(_proxy));
 
     bytes memory _callData =
-      abi.encodeWithSelector(CommonActions.exitSystemCoins.selector, address(coinJoin), _coinsToExit);
+    // REVIEW: collectSystemCoins vs. exitSystemCoins
```
Comment (wei3erHase) [inline test/scopes/ProxyUser.t.sol]: collect is: send to proxy + exit
exit is: call coinJoin.exit (coins should already be on the proxy)

File: test/simulations/pid-controller/decay_with_impulse.csv
Comment (wei3erHase) [inline test/simulations/pid-controller/decay_with_impulse.csv:1]: i don't see how simulations can be modified, please revert these changes

File: test/scopes/DirectUser.t.sol [hunk 4]
```diff
@@ -42,6 +42,20 @@ abstract contract DirectUser is BaseUser, Contracts, ScriptBase {
...
+    vm.stopPrank();
+  }
+
+  function _exitCoin(address _user, uint256 _amount) internal override {
+    vm.startPrank(_user);
+    safeEngine.approveSAFEModification(address(coinJoin));
+    coinJoin.exit(_user, _amount);
+    vm.stopPrank();
+  }
+
```
Comment (wei3erHase) [inline test/scopes/DirectUser.t.sol]: unnecessary refactor imo

File: test/simulations/pid-controller/decay_with_impulse.csv
Comment (wei3erHase) [reply test/simulations/pid-controller/decay_with_impulse.csv:1]: oh, i see that `dev` has simulations wrong, when calling `yarn test:simulation` it changes these files (addressing in #49), either way, not for this commit, plz revert

File: src/contracts/proxies/actions/BasicActions.sol (hunk 1 above)
Comment (0xJabberwock) [reply src/contracts/proxies/actions/BasicActions.sol]: ~I added it just in case. I think it's good for increasing fault tolerance.~

In the end, I got rid of it so as to keep the PR minimal.

File: src/interfaces/proxies/actions/ICommonActions.sol (hunk 2 above)
Comment (0xJabberwock) [reply src/interfaces/proxies/actions/ICommonActions.sol]: ~I think it may be misleading to input a `WAD` value at `joinSystemCoins()`, but a `RAD` value at `exitSystemCoins()`. Considering that both `coinJoin.join()` and `coinJoin.exit()` accept a `WAD` amount, it is even more confusing. What do you think?~

As the previous matter had most probably been an already addressed design decision, I assumed that what was wrong here was the documentation notes, which said `[wad]` rather than `[rad]`.

File: test/scopes/DirectUser.t.sol (hunk 4 above)
Comment (0xJabberwock) [reply test/scopes/DirectUser.t.sol]: I replicated the order found in the proxy actions. 😇

File: src/interfaces/proxies/actions/ICommonActions.sol (hunk 2 above)
Comment (0xJabberwock) [reply src/interfaces/proxies/actions/ICommonActions.sol]: As the previous matter had most probably been an already addressed design decision, I assumed that what was wrong here was the documentation notes, which said `[wad]` rather than `[rad]`.

File: test/scopes/ProxyUser.t.sol (hunk 3 above)
Comment (0xJabberwock) [reply test/scopes/ProxyUser.t.sol]: `collectTokenCollateral()`, which transfers collateral from the SAFE handler to the proxy and then to the user, now exists within the `BasicActions`.

Nonetheless, I noticed that `collectProtocolTokens()`, which transfers protocol tokens from the proxy to the user, already existed within the `DebtBidActions`.

Thus, taking into account the provided descriptions of `collect` and `exit`, I find a naming convention contradiction between the former and the latter.

File: test/e2e/E2ETest.t.sol [hunk 5]
```diff
@@ -135,7 +137,32 @@ abstract contract E2ETest is BaseUser, Base_CType, Common {
...
+    _liquidateSAFE(_cType(), address(this));
+    _setCollateralPrice(_cType(), INITIAL_PRICE);
+
+    uint256 _amountToBid = Math.wmul(DEBT_AMOUNT, LIQUIDATION_PENALTY);
+    // NOTE: getExpectedCollateralBought doesn't have a previous reference (lastReadRedemptionPrice)
+    (uint256 _expectedCollateral,) = collateralAuctionHouse[_cType()].getCollateralBought(1, _amountToBid);
+    assertLt(_expectedCollateral, COLLATERAL_AMOUNT);
+
+    _generateDebt(address(this), address(collateralJoin[_cType()]), int256(COLLATERAL_AMOUNT), int256(DEBT_AMOUNT));
+    _generateDebt(address(this), address(collateralJoin[_cType()]), int256(COLLATERAL_AMOUNT), int256(DEBT_AMOUNT));
```
Comment (0xJabberwock) [inline test/e2e/E2ETest.t.sol]: `_generateDebt()` is being called twice so as to avoid a `ERC20InsufficientBalance` reversion. Should it be called once?

Comment (wei3erHase) [reply test/e2e/E2ETest.t.sol]: how can you not generate double collateral? not following

File: test/scopes/DirectUser.t.sol [hunk 6]
```diff
@@ -148,16 +159,17 @@ abstract contract DirectUser is BaseUser, Contracts, ScriptBase {
...
     ICollateralAuctionHouse(_collateralAuctionHouse).buyCollateral(_auctionId, _amountToBid);
+    vm.stopPrank();
 
     // exit collateral
     bytes32 _cType = ICollateralAuctionHouse(_collateralAuctionHouse).collateralType();
-    uint256 _decimals = ICollateralJoin(collateralJoin[_cType]).decimals();
-    uint256 _collateralWei = _soldAmount / 10 ** (18 - _decimals);
-    ICollateralJoin(collateralJoin[_cType]).exit(_user, _collateralWei);
-    vm.stopPrank();
+    _exitCollateral(_user, address(collateralJoin[_cType]), _soldAmount);
```
Comment (wei3erHase) [inline test/scopes/DirectUser.t.sol:172]: noise

File: test/scopes/ProxyUser.t.sol [hunk 7]
```diff
@@ -80,24 +113,13 @@ abstract contract ProxyUser is BaseUser, Contracts, ScriptBase {
...
-
-  function _joinCoins(address _user, uint256 _amount) internal override {
     HaiProxy _proxy = _getProxy(_user);
 
-    vm.startPrank(_user);
-    systemCoin.approve(address(_proxy), _amount);
-
     bytes memory _callData =
-      abi.encodeWithSelector(CommonActions.joinSystemCoins.selector, address(coinJoin), _user, _amount);
+     abi.encodeWithSelector(CommonActions.exitCollateral.selector, _collateralJoin, _amount);
```
Comment (wei3erHase) [inline test/scopes/ProxyUser.t.sol]: hmmmmmmmmmmmmm 🤔

File: src/interfaces/proxies/actions/ICommonActions.sol (hunk 2 above)
Comment (wei3erHase) [reply src/interfaces/proxies/actions/ICommonActions.sol]: hmmmm and we're doing the `/ RAY` inside, it's weird, but ok, good catch fixing the comment

File: test/scopes/ProxyUser.t.sol (hunk 3 above)
Comment (wei3erHase) [reply test/scopes/ProxyUser.t.sol]: hmmm because there's no Join for the protocol token i guess, but yes, i agree that it may be an edge case to the convention

File: test/scopes/DirectUser.t.sol (hunk 4 above)
Comment (wei3erHase) [reply test/scopes/DirectUser.t.sol]: but u erased and created a lot of new lines on audited and forked code my fren

File: test/scopes/ProxyUser.t.sol [hunk 8]
```diff
@@ -229,7 +229,6 @@ abstract contract ProxyUser is BaseUser, Contracts, ScriptBase {
     uint256 _amountToBid
   ) internal override {
     HaiProxy _proxy = _getProxy(_user);
-    _joinCoins(_user, _amountToBid);
```
Comment (wei3erHase) [inline test/scopes/ProxyUser.t.sol:200]: 👍

**Review by wei3erHase — APPROVED** [review]
LGTM, a bit messy tbh, would have preferred no reorder, but checked them and they're ok 👍 
great battle! ⚔️
